### PR TITLE
Fix Cupertino route animation.

### LINKED ISSUE #48225: iOS App Freeze for ~0.5 second on Swipe Back Gesture
## Steps to Reproduce

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(Root());

class Root extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(home: First());
  }
}

class First extends StatefulWidget {
  @override
  _FirstState createState() => _FirstState();
}

class _FirstState extends State<First> {
  var tappedTimes = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('First'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: const Text('Go To Second'),
              onPressed: () {
                setState(() {
                  tappedTimes++;
                });
                Navigator.of(context).push(
                  CupertinoPageRoute(
                    builder: (BuildContext context) {
                      return Second();
                    },
                  ),
                );
              },
            ),
            InkWell(
              child: Text('I\'m tapped $tappedTimes times'),
            )
          ],
        ),
      ),
    );
  }
}

class Second extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Second'),
      ),
      body: Container(),
    );
  }
}
```
(Code is modified from https://github.com/flutter/flutter/issues/41024)

1. Click  'Go To Second'
2. Click go back button on the top left then quickly keep trying clicking 'Go To Second', 'Go To Second' button is clickable right away
3. SWIPE back then quickly keep trying clicking 'Go To Second' again, notice that screen freezes for ~0.5 seconds

https://imgur.com/a/4Ubzrj4
Here is a demo, in the first few times I clicked go back button on the top left and as you can see, I'm able to click 'Go To Second' immediately after navigating back to the first screen, however starting "I'm tapped 5 times", there is a significantly delay before 'Go To Second' button is clickable since swipe back is used.

**Target Platform:** iPhone / iPad
**Target OS version/browser:** iOS13.3
**Devices:** Reproduced in multiple devices

## Logs

```
```

```
```

```
flutter doctor                   
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[!] Android Studio (version 3.4)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
[!] VS Code (version 1.40.0)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[✓] Connected device (1 available)

```

### LINKED ISSUE #62848: iOS back button/gesture animation blocks UI
1. Create CupertinoApp, with two pages: something like ListView, and on tap on any element you should go to Page 2 with CupertinoPageScaffold and CupertinoNavigationBar.

2. Use CupertinoPageRoute from page 1 to go to page 2

3. Use back button or back gesture on page 2 AND immediately start to scroll page 1 when it appears.

Code example:
https://gist.github.com/v1z4/ea776251e5244fe4c7edb55a460e5b9e

Expected: when I see animation is finished, I should be able to scroll the page

Actual: there is a huge delay and interface is blocked.

I have checked the source, looks like the problem is here:
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/cupertino/route.dart#L774

1. Curves.fastLinearToSlowEaseIn is too slow at the end
2. droppedPageBackAnimationTime takes too much time
3. And UI is being blocked during the animation

I have an application and I read a lot of user complains about that behavior, so I have to modify it and increase animation speed, however it started to look a bit worse.

 flutter doctor -v

> 
> [✓] Flutter (Channel unknown, 1.19.0-5.0.pre.3, on Mac OS X 10.15.5 19F101, locale ru-RU)
>     • Flutter version 1.19.0-5.0.pre.3 at /Users/v1z/Work/iOS/flutter
>     • Framework revision 20be16d3af (12 дней назад), 2020-07-23 22:27:50 +0300
>     • Engine revision 60b269d898
>     • Dart version 2.9.0 (build 2.9.0-21.2.beta)
> 
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
>     • Android SDK at /Users/v1z/Library/Android/sdk
>     • Platform android-30, build-tools 30.0.0
>     • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
>     • All Android licenses accepted.
> 
> [✓] Xcode - develop for iOS and macOS (Xcode 11.5)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 11.5, Build version 11E608c
>     • CocoaPods version 1.9.3
> 
> [✓] Chrome - develop for the web
>     • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
> 
> [✓] Android Studio (version 4.0)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 47.1.2
>     • Dart plugin version 193.7361
>     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
> 
> [✓] VS Code (version 1.47.3)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.13.1
> 
> [✓] Connected device (3 available)
>     • iPhone X (mobile) • 42F80A2E-5BE6-4165-826D-BB10D0CB9072 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)
>     • Web Server (web)  • web-server                           • web-javascript • Flutter Tools
>     • Chrome (web)      • chrome                               • web-javascript • Google Chrome 84.0.4147.105

### LINKED ISSUE #73026: [CupertinoPageRoute] Gestures don't respond when clicking immediately after swiping back
**The problem is that after I use cupertionPageroute, when I swipe back to the previous page on iOS device, there is a high probability that the first time I click on the content of the previous page, nothing happens until I click twice, especially when you are halfway through the slide. Let go, and then quickly click on the previous page. The online business here is very seriously affected.**

### video

https://user-images.githubusercontent.com/23304639/103204325-11c13800-4932-11eb-93c2-a0bce2b11723.mov

### flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, 1.22.4, on macOS 11.1 20C69 darwin-x64, locale zh-Hans-CN)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
[!] Android Studio (version 4.1)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.52.1)
[✓] Connected device (2 available)

! Doctor found issues in 2 categories.

### this is the mini example

**main.dart**

```import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:trace_startup/page.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: FlatButton(
        color: Colors.blue,
        onPressed: () {
          Navigator.push(context, CupertinoPageRoute(builder: (context) {
            return SecondPage();
          }));
        },
        child: Container(
          height: double.infinity,
          width: double.infinity,
          child: Center(
            child: Text(
              'push new page',
            ),
          ),
        ),
      ),
    );
  }
}
```

**page.dart**
```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class SecondPage extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("second page"),),
      body: Center(child: Text("second page"),),
    );
  }

}
```

### PR DESCRIPTION
**Pull request by yiiim**
Fixes: #48225 
Fixes: #73026 
Fixes: #62848

After some research, I found that the issue might simply be that our animation curve is too steep, causing the animation to continue while the page visually appears very close to popping. This PR modifies the animation curve and duration after the drag gesture is released (iOS native animation duration is not related to swipe distance). I recorded a video comparing the curves. From top to bottom, the video shows the iOS native curve, the curve modified by this PR, and the original curve. The animation duration is slowed down by 10 times.

1/2

https://github.com/user-attachments/assets/77d0a782-b56d-431b-b925-8ff4e825c14a

1/4

https://github.com/user-attachments/assets/a4c50219-e86d-4cce-8a92-b266eb6260a8

forward  1/2

https://github.com/user-attachments/assets/067fffc2-203b-4686-ba4c-3b61a2c98cf8

forward  1/4

https://github.com/user-attachments/assets/c1ae938f-76ab-42f8-a832-d2d0e6c0758d

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/route.dart [hunk 1]
```diff
@@ -30,13 +30,8 @@ import 'localizations.dart';
...
 
-// An eyeballed value for the maximum time it takes for a page to animate forward
-// if the user releases a page mid swipe.
-const int _kMaxDroppedSwipePageForwardAnimationTime = 800; // Milliseconds.
-
-// The maximum time for a page to get reset to it's original position if the
-// user releases a page mid swipe.
-const int _kMaxPageBackAnimationTime = 300; // Milliseconds.
+// An value for the time it takes for a page to animate if the user releases a page mid swipe.
+const int _kDroppedSwipePageAnimationTime = 350; // Milliseconds.
```
Comment (victorsanni) [inline packages/flutter/lib/src/cupertino/route.dart:34]: Small nit: Might make it easier to read if the entire duration was defined here instead. Then the `// Milliseconds.` comment wouldn't be needed.

**Comment by SmallTownRiver** [comment]
I've been waiting for this for a long time. I need it so much.

File: packages/flutter/lib/src/cupertino/route.dart [hunk 2]
```diff
@@ -30,13 +30,8 @@ import 'localizations.dart';
...
 
-// An eyeballed value for the maximum time it takes for a page to animate forward
-// if the user releases a page mid swipe.
-const int _kMaxDroppedSwipePageForwardAnimationTime = 800; // Milliseconds.
-
-// The maximum time for a page to get reset to it's original position if the
-// user releases a page mid swipe.
-const int _kMaxPageBackAnimationTime = 300; // Milliseconds.
+// The duration for a page to animate when the user releases it mid-swipe.
+const Duration _kDroppedSwipePageAnimationDuration = Duration(milliseconds: 350);
```
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/route.dart:34]: Do we always want to use the same duration for forward and back now? The videos you have are only for if the drag is not far enough to start the pop.

**Review by victorsanni — APPROVED** [review]
LGTM pending Mitchell's comment.

File: packages/flutter/lib/src/cupertino/route.dart (hunk 2 above)
Comment (yiiim) [reply packages/flutter/lib/src/cupertino/route.dart:34]: I am certain that regardless of moving forward or backward, and regardless of the drag distance, the animation duration on iOS native is always the same. I have supplemented the video of moving forward.

**Review by MitchellGoodwin — APPROVED** [review]
LGTM!

**Comment by SmallTownRiver** [comment]
I see this PR has been merged into the mainline. When will it be released? I am looking forward to it.

**Comment by MitchellGoodwin** [comment]
It's in the beta branches as of 3.26.0-0.1.pre and up, and it will be on the next stable release. Stable releases aren't announced ahead of time, unfortunately.

https://github.com/flutter/flutter/blob/master/docs/releases/Where's-my-commit.md

**Comment by SmallTownRiver** [comment]
@MitchellGoodwin 

Thank you so much for your reply

May I ask when is the next stable version release？

**Comment by victorsanni** [comment]
Hi @yiiim, just to follow up, I wanted to ask how you were able to draw the lines and record the videos in this PR, they look really good! I'm trying to fix a similar route animation issue in https://github.com/flutter/flutter/pull/162097.

**Comment by yiiim** [comment]
@victorsanni  For iOS native, use its "Slow Animations" feature to slow down animations. For Flutter, modify the animation duration in the code. Then, I use CapCut to merge them into a single video.

**Comment by tcgumus** [comment]
I'm still having the same problem on https://github.com/flutter/flutter/issues/48225. Do i need to create a new issue for it?

**Pull request merged**